File: 2Arrays and Strings/Code/JavaScript/2-2IsSubsequence.js
/**
 * @param {string} s
 * @param {string} t
 * @return {boolean}
 */
var isSubsequence = function(s, t) {
    let i = 0, j = 0;
    while (i < s.length && j < t.length) {
        if (s[i] == t[j]) {
            i++;
        }
        
        j++;
    }
    
    return i == s.length;
};

// Test-cases
console.log(isSubsequence("ace", "abcde"));   // true

console.log(isSubsequence("aec", "abcde"));   // false

console.log(isSubsequence("abc", "ahbgdc"));  // true

console.log(isSubsequence("axc", "ahbgdc"));  // false

// Compilation steps:
// node 2-2IsSubsequence.js

// Time-complexity : O(n+m)
//   n: length of array 1
//   m: length of array 2

// Space complexity: O(1)

File: 2Arrays and Strings/Code/Python/2-2IsSubsequence.py
def isSubsequence(s: str, t: str) -> bool:
    i = j = 0
    while i < len(s) and j < len(t):
        if s[i] == t[j]:
            i += 1
        j += 1

    return i == len(s)

# test-cases
print(isSubsequence("ace", "abcde"))  # true

print(isSubsequence("aec", "abcde"))  # false

print(isSubsequence("abc", "ahbgdc")) # true

print(isSubsequence("axc", "ahbgdc")) # false

# Time-complexity: O(n+m) where
## n: length of string 1
## m: length of string 2

# Space-complexity: O(1)
